src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -193,48 +202,84 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
      * </ol>
      *
      * @param programmingExercise The programmingExercise that should be setup
-     * @param isImportedFromFile  defines if the programming exercise is imported from a file
+     * @param isImportedFromFile  defines if the programming exercise is imported from a file, if the
+     *                                exercise is imported, the build plans will not be triggered to prevent erroneous builds
      * @return The new setup exercise
      * @throws GitAPIException If something during the communication with the remote Git repository went wrong
      * @throws IOException     If the template files couldn't be read
      */
-    @Transactional // TODO: apply the transaction on a smaller scope
-    // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
     public ProgrammingExercise createProgrammingExercise(ProgrammingExercise programmingExercise, boolean isImportedFromFile) throws GitAPIException, IOException {
-        programmingExercise.generateAndSetProjectKey();
-        final User exerciseCreator = userRepository.getUser();
+        AtomicBoolean gitExceptionThrown = new AtomicBoolean(false);
 
+        final User exerciseCreator = userRepository.getUser();
         VersionControlService versionControl = versionControlService.orElseThrow();
+
+        programmingExercise.generateAndSetProjectKey();
         programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
-        programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
-        initParticipations(programmingExercise);
-        setURLsAndBuildPlanIDsForNewExercise(programmingExercise);
 
-        // Save participations to get the ids required for the webhooks
-        connectBaseParticipationsToExerciseAndSave(programmingExercise);
+        ProgrammingExercise createdExercise = transactionTemplate.execute(new TransactionCallback<ProgrammingExercise>() {
+
+            @Override
+            public ProgrammingExercise doInTransaction(@NotNull TransactionStatus status) {
+                try {
+                    programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
+                }
+                catch (GitAPIException e) {
+                    gitExceptionThrown.set(true);
```

Comment:
I see two problems with this:
1. The actual exception is lost -> At least print it
2. The rollback behavior is not triggered for this, as it only trigger if an exception occurs -> Rethrow a RuntimeException to trigger rollbacks and catch it outside the transaction block